x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/DocumentPermissions.java

```diff
@@ -46,58 +48,67 @@
  * queries are used as an additional filter.
  */
 public final class DocumentPermissions implements CacheKey {
-    // SortedSet because orders are important when they get serialised for request cache key
-    private final SortedSet<BytesReference> queries;
-    private final SortedSet<BytesReference> limitedByQueries;
-    private List<String> evaluatedQueries;
-    private List<String> evaluatedLimitedByQueries;
 
-    private static DocumentPermissions ALLOW_ALL = new DocumentPermissions();
+    @Nullable
+    private final List<Set<BytesReference>> listOfQueries;
+    @Nullable
+    private List<List<String>> listOfEvaluatedQueries;
+
+    private static final DocumentPermissions ALLOW_ALL = new DocumentPermissions();
 
     DocumentPermissions() {
-        this.queries = null;
-        this.limitedByQueries = null;
+        this.listOfQueries = null;
     }
 
     DocumentPermissions(Set<BytesReference> queries) {
-        this(queries, null);
+        if (queries == null || queries.isEmpty()) {
+            throw new IllegalArgumentException("null or empty queries not permitted");
```

Comment:
I think so, i.e. it can only be programming bug. The throwing code is carried over from existing code (the `filteredBy` static method).  In general, I also favor assertions. That said, null check is often done with `Objects.requireNonNull` which is a runtime logic. Empty check is closely related. So I don't mind them to be runtime check. I can also add `assert false` inside the `if` block so we can have both. What do you think?

Separately, I'll also further lock down the constructors by making them private. They are only used in tests and should be replaced by convenient methods.

Reply:
We synced offline and agreed to go with assertions. I also made all constructors private so that instantiation is now only possible with the static methods of `filteredBy` and `limitDocumentPermissions`.